lib/active_set/filtering/operation.rb

```diff
@@ -0,0 +1,189 @@
+# frozen_string_literal: true
+
+require_relative '../attribute_instruction'
+
+class ActiveSet
+  module Filtering
+    class Operation
+      def initialize(set, instructions_hash)
+        @set = set
+        @instructions_hash = instructions_hash
+      end
+
+      def execute
+        attribute_instructions = @instructions_hash
+                                   .flatten_keys
+                                   .map { |k, v| AttributeInstruction.new(k, v) }
+
+        activerecord_filtered_set = attribute_instructions.reduce(@set) do |set, attribute_instruction|
+          maybe_set_or_false = ActiveRecordStrategy.new(set, attribute_instruction).execute
+          next set unless maybe_set_or_false
+
+          attribute_instruction.processed = true
+          maybe_set_or_false
+        end
+
+        return activerecord_filtered_set if attribute_instructions.all?(&:processed?)
+
+        attribute_instructions.reject(&:processed?).reduce(activerecord_filtered_set) do |set, attribute_instruction|
+          maybe_set_or_false = EnumerableStrategy.new(set, attribute_instruction).execute
+          maybe_set_or_false.presence || set
+        end
+      end
+
+      def operation_instructions
+        @instructions_hash.symbolize_keys
+      end
+    end
+
+    class EnumerableStrategy
+      def initialize(set, attribute_instruction)
+        @set = set
+        @attribute_instruction = attribute_instruction
+      end
+
+      def execute
+        return false unless @set.respond_to? :select
+
+        @set.select do |item|
+          if execute_attribute_comparison_for?(item)
+            next attribute_comparison_for(item)
+          elsif execute_class_method_call_for?(item)
+            next class_method_call_for(item)
+          else
+            next false
+          end
+        end
+      end
+
+      private
+
+      def execute_attribute_comparison_for?(item)
+        attribute_item = attribute_item_for(item)
+
+        return false unless attribute_item
+        return false unless attribute_item.respond_to?(@attribute_instruction.attribute)
+        return false unless attribute_item.method(@attribute_instruction.attribute).arity <= 0
+
+        true
+      end
+
+      def execute_class_method_call_for?(item)
+        attribute_item = attribute_item_for(item)
+
+        return false unless attribute_item
+        return false unless attribute_item.class
+        return false unless attribute_item.class.respond_to?(@attribute_instruction.attribute)
+        return false if attribute_item.class.method(@attribute_instruction.attribute).arity <= 0
+
+        true
+      end
+
+      def attribute_comparison_for(item)
+        @attribute_instruction
+          .value_for(item: item)
+          .public_send(
+            @attribute_instruction.operator,
+            @attribute_instruction.value
+          )
+      end
+
+      def class_method_call_for(item)
```

Comment:
Can you give me an example of a instruction hash that would cause execution to go into this method? I'm being slow and I'm not quite seeing how this gets used (though I'm pretty sure I understand what it's doing).

Reply:
This is meant to parallel the behavior of the ActiveRecord strategy, which handles scopes. This is for handling pure-Ruby "scopes" for a data model class. The bit of the tests that touch this area of functionality would be `COVERAGE=false be rspec spec/active_set/filter_spec.rb:1088`. Here are the 6 contexts that are tested in that bit of the filter test:
```
{ datetime_between: [] }
{ string_ends_with: }
{ assoc: { datetime_between: [] } }
{ assoc: { string_ends_with: } }
{ computed_assoc: { datetime_between: [] } }
{ computed_assoc: { string_ends_with: } }
```

The two contexts for the `computed_assoc` will exercise this code. So, if you have an ActiveRecord scope defined on a model, but that model is associated with the "base" model of the collection through Ruby and not thru ActiveRecord, we need to handle that case. This code handles that case.